File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    def findCircleNum(self, M):
        """
        :type M: List[List[int]]
        :rtype: int
        """
        
        if not M: return 0
        s = len(M)
        seen = set()
        
        def dfs(p):
            for q, adj in enumerate(M[p]):
                if (adj == 1) and (q not in seen):
                    seen.add(q)
                    dfs(q)
        
        cnt = 0
        for i in range(s):
            if i not in seen: 
                dfs(i)
                cnt += 1
        
        return cnt

File: 0547-number-of-provinces/0547-number-of-provinces.java
class Solution {
    public int findCircleNum(int[][] isConnected) {
        
        int n = isConnected.length;
        boolean[] visited = new boolean[n];
        int count = 0;
        
        for(int i = 0; i < n; i++)
        {
            if(!visited[i])
            {
                dfs(isConnected, visited, i);
                count++;
            }
        }
        
        return count;
    }
    
    
    public void dfs(int[][] A, boolean[] visited, int node)
    {
        visited[node] = true;
        
        ArrayList<Integer> adjList = new ArrayList<>();
        
        for(int i = 0; i < A.length; i++)
        {
            if(A[node][i] == 1 && !visited[i])
                dfs(A, visited, i);
        }
    }
}